prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,291 @@
+{
+	FI_DBG(cache->domain->prov, FI_LOG_MR, "delete %p (len: %" PRIu64 ")\n",
+	       entry->iov.iov_base, entry->iov.iov_len);
+	cache->delete_cnt++;
+
+	util_mr_cache_process_events(cache);
+
+	if (--entry->use_cnt == 0) {
+		if (entry->cached) {
+			dlist_insert_tail(&entry->lru_entry, &cache->lru_list);
+		} else {
+			util_mr_free_entry(cache, entry);
+		}
+	}
+}
+
+static int
+util_mr_cache_create(struct ofi_mr_cache *cache, const struct iovec *iov,
+		     uint64_t access, struct ofi_mr_entry **entry)
+{
+	int ret;
+
+	FI_DBG(cache->domain->prov, FI_LOG_MR, "create %p (len: %" PRIu64 ")\n",
+	       iov->iov_base, iov->iov_len);
+
+	*entry = calloc(1, sizeof(**entry) + cache->entry_data_size);
+	if (!*entry)
+		return -FI_ENOMEM;
+
+	(*entry)->iov = *iov;
+	(*entry)->access = access;
+	(*entry)->use_cnt = 1;
+
+	ret = cache->add_region(cache, *entry);
+	if (ret) {
+		free(*entry);
+		return ret;
+	}
+
+	if (++cache->cached_cnt > cache->size) {
+		(*entry)->cached = 0;
+	} else {
+		ret = ofi_monitor_subscribe(&cache->nq, iov->iov_base, iov->iov_len,
+					    &(*entry)->subscription);
+		if (ret)
+			goto err;
+		(*entry)->subscribed = 1;
+
+		if (rbtInsert(cache->mr_tree, &(*entry)->iov, *entry)) {
+			ret = -FI_ENOMEM;
+			goto err;
+		}
+		(*entry)->cached = 1;
+	}
+
+	return 0;
+
+err:
+	util_mr_free_entry(cache, *entry);
+	return ret;
+}
+
+static int
+util_mr_cache_merge(struct ofi_mr_cache *cache, const struct fi_mr_attr *attr,
+		    RbtIterator iter, struct ofi_mr_entry **entry)
+{
+	struct iovec iov, *old_iov;
+	struct ofi_mr_entry *old_entry;
+
+	iov = *attr->mr_iov;
+	do {
+		rbtKeyValue(cache->mr_tree, iter, (void **) &old_iov,
+			    (void **) &old_entry);
+
+		FI_DBG(cache->domain->prov, FI_LOG_MR, "merge %p (len: %" PRIu64 ") "
+		       "with %p (len: %" PRIu64 ")\n",
+		       iov.iov_base, iov.iov_len,
+		       old_entry->iov.iov_base, old_entry->iov.iov_len);
+
+		iov.iov_len = ((uintptr_t)
+			MAX(ofi_iov_end(&iov), ofi_iov_end(old_iov))) -
+			((uintptr_t) MIN(iov.iov_base, old_iov->iov_base));
+		iov.iov_base = MIN(iov.iov_base, old_iov->iov_base);
+
+		rbtErase(cache->mr_tree, iter);
+		old_entry->cached = 0;
+		if (old_entry->use_cnt == 0) {
+			dlist_remove(&old_entry->lru_entry);
+			util_mr_free_entry(cache, old_entry);
+		}
+
+	} while ((iter = rbtFind(cache->mr_tree, &iov)));
+
+	return util_mr_cache_create(cache, &iov, attr->access, entry);
+}
+
+int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *attr,
+			struct ofi_mr_entry **entry)
+{
+	RbtIterator iter;
+	struct iovec *iov;
+
+	util_mr_cache_process_events(cache);
+
+	assert(attr->iov_count == 1);
+	FI_DBG(cache->domain->prov, FI_LOG_MR, "search %p (len: %" PRIu64 ")\n",
+	       attr->mr_iov->iov_base, attr->mr_iov->iov_len);
+	cache->search_cnt++;
+
+	while ((cache->cached_cnt >= cache->size) && ofi_mr_cache_flush(cache))
+		;
+
+	iter = rbtFind(cache->mr_tree, (void *) attr->mr_iov);
```

Comment:
Is this cache going to deal with access permissions as well when attempting to find a registration?